File: Intervals-Prefix-Sum/Python/238_Product_of_Array_Except_Self.py
# LeetCode 238 - Product of Array Except Self
# Pattern: Prefix & Suffix Products
# Time: O(n)
# Space: O(1) Extra Space (excluding output array)

from typing import List

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:

        n = len(nums)
        answer = [0] * n

        # Store prefix products
        answer[0] = 1
        for i in range(1, n):
            answer[i] = answer[i - 1] * nums[i - 1]

        # Multiply with suffix products
        right_product = 1
        for i in range(n - 1, -1, -1):
            answer[i] *= right_product
            right_product *= nums[i]

        return answer

File: Intervals-Prefix-Sum/JavaScript/238_Product_of_Array_Except_Self.js
// LeetCode 238 - Product of Array Except Self
// Pattern: Prefix & Suffix Products
// Time: O(n)
// Space: O(1) Extra Space (excluding output array)

var productExceptSelf = function(nums) {

    const n = nums.length;
    const answer = new Array(n);

    // Store prefix products
    answer[0] = 1;
    for (let i = 1; i < n; i++) {
        answer[i] = answer[i - 1] * nums[i - 1];
    }

    // Multiply with suffix products
    let rightProduct = 1;

    for (let i = n - 1; i >= 0; i--) {
        answer[i] *= rightProduct;
        rightProduct *= nums[i];
    }

    return answer;
};
